File: desar_manufacturing/desar_manufacturing/page/desar_supervisor/desar_supervisor.js
frappe.pages["desar-supervisor"].on_page_load = function(wrapper) {
    frappe.ui.make_app_page({
        parent: wrapper,
        title: "Production Status",
        single_column: true,
    });

    $(wrapper).find(".layout-main").html(frappe.render_template("desar_supervisor", {}));
    new DESARSupervisor(wrapper).init();
};

class DESARSupervisor {
    constructor(wrapper) {
        this.$w = $(wrapper);
    }

    init() {
        this.$w.find(".desar-sup-refresh").on("click", () => this.load());
        this.load();
        // Auto-refresh every 60 seconds
        this._timer = setInterval(() => this.load(), 60000);
    }

    load() {
        const $list = this.$w.find(".desar-sup-list").html("");
        const $loading = this.$w.find(".desar-sup-loading").show();
        const $empty = this.$w.find(".desar-sup-empty").hide();

        frappe.call({
            method: "desar_manufacturing.desar_manufacturing.page.desar_supervisor.desar_supervisor.get_production_status",
            callback: (r) => {
                $loading.hide();
                const orders = r.message || [];
                if (!orders.length) { $empty.show(); return; }
                orders.forEach(o => $list.append(this._render(o)));
            },
            error: () => $loading.text("Error loading. Click Refresh.")
        });
    }

    _render(order) {
        const pct = order.total ? Math.round(order.completed / order.total * 100) : 0;
        const color = pct === 100 ? "#27ae60" : pct > 0 ? "#2E86AB" : "#888";

        const stages = (order.work_orders || [])
            .sort((a, b) => (a.stage_name > b.stage_name ? 1 : -1))
            .map(wo => {
                const icon = wo.status === "Completed" ? "✅" :
                             wo.status === "In Process" ? "⏳" : "○";
                const url = `/app/work-order/${wo.name}`;
                return `<span class="desar-stage-pill ${wo.status === 'Completed' ? 'done' : wo.status === 'In Process' ? 'active' : ''}">
                    ${icon} <a href="${url}">${wo.stage_name || wo.production_item}</a>
                </span>`;
            }).join("");

        const so_link = order.sales_order
            ? `<a href="/app/sales-order/${order.sales_order}">${order.sales_order}</a>`
            : "—";

        return `
        <div class="desar-sup-card">
            <div class="desar-sup-card-header">
                <div>
                    <strong>${order.article_name || order.design_no || "Unknown"}</strong>
                    <span class="text-muted"> · ${so_link}</span>
                    <span class="text-muted"> · ${order.planned_date || ""}</span>
                </div>
                <div class="desar-progress-wrap">
                    <div class="desar-progress-bar">
                        <div class="desar-progress-fill" style="width:${pct}%;background:${color}"></div>
                    </div>
                    <span class="desar-pct" style="color:${color}">${pct}%</span>
                </div>
            </div>
            <div class="desar-stages">${stages}</div>
        </div>`;
    }
}

File: desar_manufacturing/desar_manufacturing/page/desar_supervisor/desar_supervisor.py
"""
DESAR Supervisor — Production Status Page Backend
Shows all active production orders with stage completion status.
"""
import frappe
from frappe.utils import today

from desar_manufacturing.constants import SUPERVISOR_ROLES


@frappe.whitelist()
def get_production_status():
    """
    Returns active production orders grouped by Production Plan.
    Each group shows all 4 WOs with their status.
    """
    frappe.only_for(SUPERVISOR_ROLES)
    # Get active Work Orders from last 30 days
    wos = frappe.get_all(
        "Work Order",
        filters={
            "docstatus": 1,
            "status": ["in", ["Not Started", "In Process", "Completed"]],
            "production_plan": ["!=", ""],
        },
        fields=[
            "name", "production_item", "qty", "status",
            "production_plan", "sales_order",
            "custom_design_no", "custom_article_name", "custom_design_master",
            "planned_start_date",
        ],
        order_by="planned_start_date asc",
        limit=200,
    )

    # Group by Production Plan
    groups = {}
    for wo in wos:
        pp = wo.production_plan
        if pp not in groups:
            so = wo.sales_order or ""
            groups[pp] = {
                "production_plan": pp,
                "sales_order": so,
                "design_no": wo.custom_design_no or "",
                "article_name": wo.custom_article_name or "",
                "planned_date": str(wo.planned_start_date or ""),
                "work_orders": [],
                "total": 0,
                "completed": 0,
            }

        # Get stage name
        stage_name = ""
        if wo.custom_design_master and wo.production_item:
            stage_name = frappe.db.get_value(
                "DESAR Stage Configuration",
                {"parent": wo.custom_design_master, "output_item": wo.production_item},
                "stage_name"
            ) or wo.production_item

        groups[pp]["work_orders"].append({
            "name": wo.name,
            "production_item": wo.production_item,
            "stage_name": stage_name,
            "qty": wo.qty,
            "status": wo.status,
        })
        groups[pp]["total"] += 1
        if wo.status == "Completed":
            groups[pp]["completed"] += 1

    # Sort — in-progress first, then not started, then completed
    result = list(groups.values())
    def sort_key(g):
        if g["completed"] < g["total"] and g["completed"] > 0:
            return 0  # In progress
        elif g["completed"] == 0:
            return 1  # Not started
        else:
            return 2  # Completed
    result.sort(key=sort_key)

    return result
